File: app_alertas/lib/core/report_visibility.dart
import 'package:app_alertas/models/alert_model.dart';

/// Reportes sin actividad reciente no se muestran en el mapa.
const reportMapStaleDuration = Duration(hours: 2);

DateTime alertLastActivityAt(AlertModel alert) {
  final created = alert.createdAt ?? DateTime.fromMillisecondsSinceEpoch(0);
  final updated = alert.updatedAt ?? created;
  if (updated.isAfter(created)) return updated;
  return created;
}

bool isAlertVisibleOnMap(AlertModel alert, {DateTime? now}) {
  final reference = now ?? DateTime.now();
  final lastActivity = alertLastActivityAt(alert);
  return reference.difference(lastActivity) <= reportMapStaleDuration;
}

List<AlertModel> filterAlertsForMap(List<AlertModel> alerts) {
  return alerts.where(isAlertVisibleOnMap).toList();
}

File: web_alertas/src/lib/report-visibility.ts
import type { Report } from "@/domain/types";

/** Reportes sin actividad reciente no se muestran en el mapa. */
export const REPORT_MAP_STALE_MS = 2 * 60 * 60 * 1000;

export function getReportLastActivityAt(report: Report): Date {
  const created = new Date(report.created_at).getTime();
  const updated = report.updated_at
    ? new Date(report.updated_at).getTime()
    : created;

  const imageTimes = (report.images ?? [])
    .map((img) => (img.uploaded_at ? new Date(img.uploaded_at).getTime() : 0))
    .filter((t) => t > 0);

  const latestImage = imageTimes.length ? Math.max(...imageTimes) : 0;
  return new Date(Math.max(created, updated, latestImage));
}

export function isReportVisibleOnMap(report: Report, now = Date.now()): boolean {
  const lastActivity = getReportLastActivityAt(report).getTime();
  return now - lastActivity <= REPORT_MAP_STALE_MS;
}

export function filterReportsForMap(reports: Report[]): Report[] {
  return reports.filter(isReportVisibleOnMap);
}
